tools/base/argsfilter/argsfilter.c

```diff
@@ -0,0 +1,221 @@
+//
+// Redistribution  and  use  in source  and  binary  forms,  with  or  without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of  source code  must retain the  above copyright notice,
+//   this list of conditions and the following disclaimer.
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+// * Neither the name  of Intel Corporation  nor the names of its contributors
+//   may be used to  endorse or promote  products derived  from this  software
+//   without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+#include "argsfilter.h"
+#include <getopt.h>
+#include <stdlib.h>
+#include <string.h>
+#ifdef _WIN32
+#define EX_OK 0
+#define EX_USAGE (-1)
+#define EX_SOFTWARE (-2)
+#else
+#include <sysexits.h>
+#endif
+
+#define RETURN_ON_ERR(res, desc)                                               \
+	do {                                                                   \
+		if ((res) != FPGA_OK) {                                        \
+			optind = 1;                                            \
+			opterr = old_opterr;                                   \
+			fprintf(stderr, "Error %s: %s\n", (desc),              \
+				fpgaErrStr(res));                              \
+			return EX_SOFTWARE;                                    \
+		}                                                              \
+	} while (0)
+
+int set_properties_from_args(fpga_properties filter, fpga_result *result,
+			     int *argc, char *argv[])
+{
+	// prefix the short options with '-' so that unrecognized options are
+	// ignored
+	const char *short_opts = "-:B:D:F:S:";
+	struct option longopts[] = {
+		{"bus", required_argument, NULL, 'B'},
+		{"device", required_argument, NULL, 'D'},
+		{"function", required_argument, NULL, 'F'},
+		{"socket-id", required_argument, NULL, 'S'},
+		{"segment", required_argument, NULL, 0xe},
+		{0, 0, 0, 0},
+	};
+	int supported_options = sizeof(longopts) / sizeof(longopts[0]) - 1;
```

Comment:
I think this may get flagged by static analysis. You might want to use a #define